iquip/apps/builder.py

```diff
@@ -0,0 +1,419 @@
+"""App module for editting the build arguments and submitting the experiment."""
+
+import json
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
+
+import requests
+from PyQt5.QtCore import QObject, Qt, QThread, pyqtSignal, pyqtSlot
+from PyQt5.QtWidgets import (
+    QCheckBox, QComboBox, QDoubleSpinBox, QHBoxLayout, QLabel, QLineEdit, QListWidget,
+    QListWidgetItem, QPushButton, QVBoxLayout, QWidget
+)
+
+import qiwis
+from iquip.protocols import ExperimentInfo
+
+class _BaseEntry(QWidget):
+    """Base class for all argument entries.
+
+    This is a wrapper of each entry.
+    In each subclass, value() must be implemented to return the selected value.
+
+    Attributes:
+        name: The argument name.
+        nameLabel: The label for showing the argument name.
+    """
+
+    def __init__(self, name: str, parent: Optional[QWidget] = None):
+        """Extended.
+        
+        Args:
+            name: See the attributes section in _BaseEntry.
+        """
+        super().__init__(parent=parent)
+        self.name = name
+        # widgets
+        self.nameLabel = QLabel(name, self)
+        # layout
+        self.layout = QHBoxLayout(self)
+        self.layout.addWidget(self.nameLabel)
+        self.setLayout(self.layout)
+
+    def value(self) -> Any:
+        """Returns the entered or selected value.
+        
+        This must be overridden in the subclass.
+        """
+        raise NotImplementedError
+
+
+class _BooleanEntry(_BaseEntry):
+    """Entry class for a boolean value."""
+
+    def __init__(self, name: str, default: bool = False, parent: Optional[QWidget] = None):
+        """Extended.
+        
+        Args:
+            default: The default value. If it does not exist, it is set to False.
+        """
+        super().__init__(name, parent=parent)
+        # widgets
+        self.checkBox = QCheckBox(self)
+        self.checkBox.setChecked(default)
+        # layout
+        self.layout.addWidget(self.checkBox)
+
+    def value(self) -> bool:
+        """Overridden.
+        
+        Returns the status of the checkBox.
+        """
+        return self.checkBox.isChecked()
+
+
+class _EnumerationEntry(_BaseEntry):
+    """Entry class for an enumeration value."""
+
+    def __init__(
+        self,
+        name: str,
+        choices: List[str],
```

Comment:
Please take abstract type for the parameters, e.g., `Iterable`.
However, of course, in this case, you used `if choices` so `Iterable` is not appplicable.
Nevertheless, what happens if the `choices` is empty? Is it a valid situation?
I think it should be considered as an invalid input.

Reply:
The case the `choices` is empty never happens because in that case fetching the build arguments raises an exception.
Nevertheless if you think we should handle it, I will add it.

Reply:
Also, the proxy server always offers the `choices` with a list type and should be the type modified to `Iterable`?

Reply:
> The case the `choices` is empty never happens because in that case fetching the build arguments raises an exception.
> Nevertheless if you think we should handle it, I will add it.

No, my point was: "is `if choices` check necessary?"
You can assume that `choices` are always not empty as you said.
Because of `if choices`, you cannot use `Iterable` type annotation.

However, as you said, I think it is okay to use `List` since this is an internal class.

Conclusion: You can keep or change the current code as you want.

Reply:
Oh, I see. Then, I will keep it as it is.

Reply:
Oh, I'm sorry that I said the wrong thing about `choices`..

First, `choices` is always passed, i.e. it is guaranteed that it isn't `None`.
However, it may be an empty list and the experiment __CANNOT__ be successful since the argument cannot be assigned to any value.
But I think this PR isn't responsible for this problem, so I just ignore the case.

Reply:
I don't understand what you said.
1. Why do you care if `choices` is `None`? You didn't set the default value for it.
2. Do you mean that an empty list can be given to `choices`?

Please explain more about this sentence:

> However, it may be an empty list and the experiment CANNOT be successful since the argument cannot be assigned to any value.

Reply:
Okay! I said it wrong at first, so I corrected it below.

The `choices` is always given so I didn't set its default value.
But the problem is when it is set as an empty list.
In that case, an user cannot set it to any value because there is no allowed candidate.
Once the user submits it, it causes an exception that there is an empty argument.

Reply:
Okay, then `choices` here is always not empty, right?

Reply:
Nope! It may be an empty list! (Is it right that 'empty` means an emtpy list?)
If an user set `choices` as an empty list in code, the builder is open well but the execution would be failed.

Reply:
Oh, now I understand. Okay then, please create an issue for handling an empty enumeration value.